### PR TITLE
Tips

### PR DESCRIPTION
**Pull request by DynamicSquid**
I added a version command if that's helpful :)

Also just some things I noticed

---

For your lexer, instead of getting all the tokens at once, you could also get them when you need them, like this:

```rs
fn parser() {
    tok = lexer.eat(); // get next token
    if tok.type == TokType::IF {
        // get next token; if next token is not an open bracket, throw error
        tok = lexer.eat(TokType::OPEN_BRACKET, "expected open bracket after 'if' keyword");
    }
}
```

It doesn't really matter which way you do it. I kinda like the second way thought since you have more control over the lexing process, and it's much easier to deal with syntax errors.

---

For your parser, you have a lot of duplicate code with your functions: `comp`, `add`, `mult`, `unary`. I would just suggest having `unary` and `binary`.

Same thing with `types.rs`, there's a bunch of duplicate code. I'm not sure how it works in Rust, but you could try to have a generic binary expression function, and just pass in a lambda like this:

```rs
pub fn evaluate_binary_expression(
    &self,
    other: &Type,
    operation:   &dyn Fn(i32, i32) -> TResult,        // lambda function
    div_by_zero: bool,                                // if false, and division by 0 is encountered, throw error
    op_name:     &str                                 // name of the operator to display in the error message
) -> TResult {

        if let (Self::Float(a), Self::Float(b)) = (self, other) {
            if !div_by_zero && *b == 0f32 {
                return Err("Division by 0.".into());
            }

            return Ok(Self::Float(operation(a, b)));
        }

        // not sure how str concat work in rust :/
        Err("Operator '" + op + "' can only be applied to numbers.".into())
}
```

Hope it helps :D

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -25,6 +25,11 @@ fn main() {
         process::exit(1);
     }
 
+    if args[0].eq("--version") {
+        println!("Europa Lang v0.0.0");
+        process::exit(1);
```
Comment (19wintersp) [inline src/main.rs:30]: Why do you exit with code 1?

Comment (19wintersp) [reply src/main.rs:30]: (most programs will exit normally when you run --version)

image

**Comment by cursorweb** [comment]
1. lexer -- the solution you proposed doesn't exactly show how you can show the expression part: `if <expression> {`. Indeed, it is useful, but it also confuses the parsing step with the parser, which I would like to keep distinct, unless there really is a better way to do it -- I'm all ears!
2. Yeah, I might make a macro to handle that, but I have to keep the functions distinct to have operator precedence.
3. Good idea, I'll have to read up on lambdas. Thanks!

File: src/main.rs (hunk 1 above)
Comment (cursorweb) [reply src/main.rs:30]: let's also add the `-v` flag, common in windows cli.

Comment (19wintersp) [reply src/main.rs:30]: @cursorweb Would it be "-v" or "-V"? Either one can be used to mean "version" or "verbose", and there isn't too much consensus on it.

**Review by cursorweb — CHANGES_REQUESTED** [review]
The code won't work yet because `--version` is actually on index 1. Index 0 is reserved for the executable.

Secondly, I recommend also moving it above the argument check, and also printing rust compiler version.

File: src/main.rs (hunk 1 above)
Comment (19wintersp) [reply src/main.rs:30]: I've just noticed this doesn't actually work either. `argv[0]` is the program name, and you've also asserted that the amount of arguments must be 2. Should we bring in a argument parser?

**Review by 19wintersp — CHANGES_REQUESTED** [review]
I think this would be better if we just started using an argument parser, it'll be better in the long run.

**Comment by cursorweb** [comment]
Thank you! I will create a new issue with all the ideas you discussed.

File: src/main.rs [hunk 2]
```diff
@@ -20,6 +20,11 @@ use interpreter::Interpreter;
 fn main() {
     let args: Vec<String> = env::args().collect();
 
+    if args[1].eq("--version") || args[1].eq("-v") {
+        println!("Europa Lang v0.0.0"); // todo: rust compiler version
```
Comment (19wintersp) [inline src/main.rs:24]: You can use Cargo env's to do this.
```suggestion
        println!("Europa Lang v{}", env!("CARGO_PKG_VERSION"));
```

File: src/main.rs (hunk 1 above)
Comment (DynamicSquid) [reply src/main.rs]: That was just meant to be an example

Comment (19wintersp) [reply src/main.rs]: @DynamicSquid What? So why did you create a pull request for it?

Comment (DynamicSquid) [reply src/main.rs]: Well, the main focus of the PR was the tips - I never expected you to merge it or anything

Comment (19wintersp) [reply src/main.rs]: @DynamicSquid Well then create an issue? PR's are for merging code, don't make one if you don't want it to be merged...

Comment (DynamicSquid) [reply src/main.rs]: But I was planning to edit some code to show you how to implement tip, but it turns out most of it was just explaining. But that doesn't, let's focus on the new Issue instead :)

File: src/main.rs (hunk 2 above)
Comment (cursorweb) [reply src/main.rs:24]: alright, thank you!

**Comment by cursorweb** [comment]
> I think this would be better if we just started using an argument parser, it'll be better in the long run.

yeah, true

**Pull request merged**